app/forms/api/profile_creation_form.rb

```diff
@@ -0,0 +1,152 @@
+module Api
+  class ProfileCreationForm
+    include ActiveModel::Model
+
+    validate :valid_jwt
+    validate :valid_user
+    validate :valid_password
+
+    attr_reader :password, :user_bundle
+    attr_reader :user_session, :service_provider
+    attr_reader :profile, :gpo_otp
+
+    def initialize(password:, jwt:, user_session:, service_provider: nil)
+      @password = password
+      @jwt = jwt
+      @user_session = user_session
+      @service_provider = service_provider
+      set_idv_session
+    end
+
+    def submit
+      @form_valid = valid?
+
+      if form_valid?
+        create_profile
+        cache_encrypted_pii
+        complete_session
+      end
+
+      response = FormResponse.new(
+        success: form_valid?,
+        errors: errors.to_hash,
+        extra: extra_attributes,
+      )
+      [response, personal_key]
+    end
+
+    private
+
+    attr_reader :jwt
+
+    def create_profile
+      profile_maker = build_profile_maker
+      profile = profile_maker.save_profile
+      @profile = profile
+      session[:pii] = profile_maker.pii_attributes
+      session[:profile_id] = profile.id
+      session[:personal_key] = profile.personal_key
+    end
+
+    def cache_encrypted_pii
+      cacher = Pii::Cacher.new(user, session)
+      cacher.save(password, profile)
+    end
+
+    def complete_session
+      complete_profile if phone_confirmed?
+      create_gpo_entry if user_bundle.gpo_address_verification?
+    end
+
+    def phone_confirmed?
+      user_bundle.vendor_phone_confirmation? && user_bundle.user_phone_confirmation?
+    end
+
+    def complete_profile
+      user.pending_profile&.activate
+      move_pii_to_user_session
+    end
+
+    def move_pii_to_user_session
+      return if session[:decrypted_pii].blank?
+      user_session[:decrypted_pii] = session.delete(:decrypted_pii)
+    end
+
+    def create_gpo_entry
+      move_pii_to_user_session
+      if session[:pii].is_a?(String)
+        session[:pii] = Pii::Attributes.new_from_json(user_session[:decrypted_pii])
```

Comment:
```suggestion
        session[:pii] = Pii::Cacher.new(current_user, user_session).fetch
```

Ah looks like I missed this in https://github.com/18F/identity-idp/pull/6054, but we can use it here all the same 🙂